src/autopas/containers/verletClusterLists/traversals/VerletClustersColoringTraversal.h

```diff
@@ -0,0 +1,169 @@
+    }
+
+    auto loopBody = [this, gridsPerColoringCell](unsigned long x, unsigned long y, unsigned long z) {
+      processColorCell(x, y, z, gridsPerColoringCell);
+    };
+
+    const auto end = ArrayMath::sub(coloringCellsPerDim, {0ul, 0ul, 0ul});
+    // We are only doing a 2D coloring.
+    if (end[2] != 1) {
+      autopas::utils::ExceptionHandler::exception("Coloring should only be 2D, not in z-direction!");
+    }
+
+    // localStride is necessary because stride is constexpr and cTraversal() wants a const &
+    auto localStride = stride;
+    this->cTraversal(std::forward<decltype(loopBody)>(loopBody), end, localStride);
+  }
+
+ private:
+  PairwiseFunctor *_functor;
+};
+
+template <class ParticleCell, class PairwiseFunctor, DataLayoutOption dataLayout, bool useNewton3>
+void VerletClustersColoringTraversal<ParticleCell, PairwiseFunctor, dataLayout, useNewton3>::processColorCell(
+    unsigned long xColorCell, unsigned long yColorCell, unsigned long zColorCell, int gridsPerColoringCell) {
+  // We are only doing a 2D coloring.
+  if (zColorCell != 0) {
+    autopas::utils::ExceptionHandler::exception("Coloring should only be 2D, not in z-direction!");
+  }
+
+  auto &clusterList = *VerletClustersTraversalInterface<Particle>::_verletClusterLists;
+  const auto cellsPerDim = clusterList.getCellsPerDimension();
+  auto &grids = clusterList.getGrids();
+  const auto &neighborLists = clusterList.getNeighborLists();
+  auto clusterSize = clusterList.getClusterSize();
+
+  for (int yInner = 0; yInner < gridsPerColoringCell; yInner++) {
+    for (int xInner = 0; xInner < gridsPerColoringCell; xInner++) {
+      unsigned long y = yColorCell * gridsPerColoringCell + yInner;
+      unsigned long x = xColorCell * gridsPerColoringCell + xInner;
+
+      // Not every coloring cell has to have gridsPerColoringCell grids in every direction.
+      if (x >= cellsPerDim[0] || y >= cellsPerDim[1]) {
+        continue;
+      }
+      unsigned long gridIndex1D = VerletClusterMaths::index1D(x, y, cellsPerDim);
+
+      auto &currentGrid = grids[gridIndex1D];
+      auto numClusters = currentGrid.numParticles() / clusterSize;
+      for (unsigned long currentCluster = 0; currentCluster < numClusters; currentCluster++) {
+        const auto &clusterNeighborList = neighborLists.at(gridIndex1D).at(currentCluster);
+        Particle *clusterStart = &currentGrid[currentCluster * clusterSize];
+        for (auto neighborClusterStart : clusterNeighborList) {
+          traverseClusterPair(clusterStart, neighborClusterStart, clusterSize);
+        }
+      }
+    }
+  }
+}
+
+template <class ParticleCell, class PairwiseFunctor, DataLayoutOption dataLayout, bool useNewton3>
+void VerletClustersColoringTraversal<ParticleCell, PairwiseFunctor, dataLayout, useNewton3>::traverseClusterPair(
+    Particle *clusterStart, Particle *neighborClusterStart, int clusterSize) {
+  if (neighborClusterStart == clusterStart) {
```

Comment:
What code duplication are you talking about? 
In #298, I branch here to call either the AoS or the SoA implementation, so in the end, this will be in another method. I would leave it as it is because of that.

Reply:
I meant the for loop heads since at least the outer loops are in both branches exactly the same. Also, a comment would be nice why this if is necessary (s.th. like "if true cluster interacts with itself")

Reply:
Ah, now I understand what you mean, sorry!

I changed the implementation a bit and pushed it. You can take a look if you think it is nicer now. In the end, one could argue that the code would be prettier when each cluster does not save itself in its neighbor list. This is something I considered to change, but since it is not that important and you approved how it is currently, I didn't do it.